core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java

```diff
@@ -0,0 +1,247 @@
+ * constructor arguments are first. On this platform with the same microbenchmarks just creating the XContentParser is around 900
+ * microseconds and using {#linkplain ObjectParser} directly adds another 300 or so microseconds. In the best case
+ * {@linkplain ObjectSpooler} allocates two additional objects per parse compared to {#linkplain ObjectParser}. In the worst case it
+ * allocates <code>2 + 2 + 2 * param_count</code> objects per parse. If this overhead is too much for you then feel free to have
+ * ObjectParser parse a secondary object and have that one call the target object's constructor.
+ */
+public final class ObjectSpooler<Value, Context> extends AbstractObjectParser<Value, Context> {
+    private final List<ParseField> constructorArgNames = new ArrayList<>();
+    private final ObjectParser<Spool, Context> objectParser;
+    private final Function<Object[], Value> builder;
+
+    /**
+     * Build the spooler.
+     * 
+     * @param name The name given to the spooler's ObjectParser for error identification. Use what you'd use if the object worked with
+     *        ObjectParser
+     * @param builder A function that builds the object from an array of Objects. Declare this inline with the spooler, casting the elements
+     *        of the array to the arguments so they work with your favorite constructor. The objects in the array will be in the same order
+     *        that you declared the {{@link #constructorArg(String)}s and none will be null. If any of the constructor arguments aren't
+     *        defined in the XContent then parsing will throw an error. We use an array here rather than a Map to save on allocations.
+     */
+    public ObjectSpooler(String name, Function<Object[], Value> builder) {
+        objectParser = new ObjectParser<>(name);
+        this.builder = builder;
+    }
+
+    /**
+     * Call this to do the actual parsing. This implements {@link BiFunction} for conveniently integrating with ObjectParser.
+     */
+    @Override
+    public Value apply(XContentParser parser, Context context) {
+        try {
+            return objectParser.parse(parser, new Spool(parser), context).finish();
+        } catch (IOException e) {
+            throw new ParsingException(parser.getTokenLocation(), "[" + objectParser.getName()  + "] failed to parse object", e);
+        }
+    }
+
+    /**
+     * Pass the {@linkplain BiConsumer} this returns to {@link #parser()}'s declare methods to declare a constructor argument. See this
+     * class's javadoc for an example. The order in which these are declared matters: it is the order that they come in the array passed to
+     * {@link #builder} and the order that missing arguments are reported to the user if any are missing. When all of these parameters are
+     * parsed from the {@linkplain XContentParser} the target object is immediately built.
+     */
+    public <FieldT> BiConsumer<Value, FieldT> constructorArg() {
```

Comment:
I like returning BiConsumer here because it keeps ConstructorArg entirely hidden - it is a rather dirty implementation detail and I'd prefer it not leak into other stuff in the code so that if we clean it up we have an easier time of it. BTW, I just pushed a patch that removes the whole class in favor of a singleton BiConsumer so there isn't even a class to return in that case!